File: _dsa_cpp/_1_data_structure/_3_stack_and_queue/_20_stock_span_problem.cpp
// https://leetcode.com/problems/online-stock-span/
// https://practice.geeksforgeeks.org/problems/stock-span-problem-1587115621/1
// https://www.geeksforgeeks.org/the-stock-span-problem/

class StockSpanner
{
    stack<pair<int, int>> st;
    int index = -1;

public:
    StockSpanner()
    {
    }

    int next(int price)
    {
        index += 1;
        while (!st.empty() && st.top().second <= price) st.pop();

        if (st.empty())
        {
            st.push({index, price});
            return index + 1;
        }
        
        int result = st.top().first;
        st.push({index, price});
        return index - result;
    }
};

//   Time complexity: O(N)
//  Space complexity:O(N)


// next greater element to left approach (by me)
vector<int> span(vector<int> arr) {
    int n = arr.size();
    stack<pair<int, int> > st;
    vector<int> v;

    for(int i=0;i<n;i++) {
        while(!st.empty() && st.top().first <= arr[i]) st.pop();
        if(!st.empty()) v.push_back(st.top().second);
        else v.push_back(-1);

        st.push(make_pair(arr[i], i));
    }

    for(int i=0;i<n;i++) {
        v[i] = i-v[i];
    }

    return v;
}

int main()
{
    vector<int> v {100, 80, 60, 70, 60, 75, 85};
    vector<int> res = span(v);
    for(int i=0;i<res.size();i++) {
        cout << res[i] << " ";
    }
}

File: _1_data_structures_go/_3_stack_queue/20-stock-span-problem.go
// https://leetcode.com/problems/online-stock-span/
// https://practice.geeksforgeeks.org/problems/stock-span-problem-1587115621/1
// https://www.geeksforgeeks.org/the-stock-span-problem/

package main

import "fmt"

type Pair struct {
	index int
	price int
}

type StockSpanner struct {
	stack []Pair
	index int
}

/**************** CONSTRUCTOR ****************/

func Constructor() StockSpanner {
	return StockSpanner{
		stack: []Pair{},
		index: -1,
	}
}

/**************** NEXT ****************/

func (s *StockSpanner) Next(price int) int {

	s.index++

	for len(s.stack) > 0 &&
		s.stack[len(s.stack)-1].price <= price {

		s.stack = s.stack[:len(s.stack)-1]
	}

	if len(s.stack) == 0 {
		s.stack = append(s.stack, Pair{s.index, price})
		return s.index + 1
	}

	prevIndex := s.stack[len(s.stack)-1].index
	s.stack = append(s.stack, Pair{s.index, price})

	return s.index - prevIndex
}
//   Time complexity: O(N)
//  Space complexity:O(N)

/**************** TEST ****************/

func main() {

	spanner := Constructor()

	fmt.Println(spanner.Next(100)) // 1
	fmt.Println(spanner.Next(80))  // 1
	fmt.Println(spanner.Next(60))  // 1
	fmt.Println(spanner.Next(70))  // 2
	fmt.Println(spanner.Next(60))  // 1
	fmt.Println(spanner.Next(75))  // 4
	fmt.Println(spanner.Next(85))  // 6
}